### ISSUE TITLE
Could `set` accept a function?

### ISSUE THREAD

**Issue by tomekwi** [tjmehta/101#49]
Is it possible to bend `set` to set properties through a function? I mean like:

```js
var markdown = require("marked");

[{string: "..."}, {string: "..."}].map(compose
  ( set("html", markdown)
  , pluck("string")
  ));
```

**Comment by tomekwi** [comment]
Maybe `set(markdown, "html")` – that should be backwards compatible.

**Comment by tomekwi** [comment]
I'll be happy to code it if you give the green light @tjmehta.

**Comment by tjmehta** [comment]
IMO overloading this function make it's less elegant and easy to understand.

It's easy to get carried away with overloading functions with functional behavior.
Sometimes writing out the implementation is easier to follow in this case::
```js
var markdown = require("marked");

[{string: "..."}, {string: "..."}].map(function (obj) {
  obj.html = markdown(obj.string)
  return obj;
});
```

What do you think?

**Comment by tomekwi** [comment]
Well yeah, that's what I did in the end.

Or rather, loving the immutability of `set`:
```js
... .map(function (obj) {
  return set(obj, "html", markdown(obj.string));
  };
```

I've just had the impression that `set` could be more useful when mapping. At the moment its only use seems to be setting some property to the same value in all items. Just what you've shown in the example.

**Comment by tomekwi** [comment]
But I must admit it seems less elegant, the way I proposed.

How about making it an optional `options` argument at the beginning? Like
```js
set({through: markdown, keypaths: false}, "html")
```

**Comment by tjmehta** [comment]
I definitely like the idea. I even considered just assuming a a function as a value in set is a transform function. What about a separate method all-together, like setMap or something?

**Comment by tomekwi** [comment]
Well, as you wish :)

I guess I'd like the `options` more as it's more concise – but it does bring its own problems along as well.

**Comment by tjmehta** [comment]
Yeah, I am not a fan of options objects, unless it's a last resort for a powerful function. I think rigidity of arguments leads to less usage confusion, and a simpler implementation.

**Comment by tomekwi** [comment]
I'll do it just as you wish :)

Now for the name, I'd go for something more generic. Because `.map` may not be the only use. Like  `setThrough` or `setVia`.

What do you think?

**Comment by tjmehta** [comment]
How about key-map or keypath-map?

[....].map(keypathMap({
  'foo.bar': compose(markdown, pluck('string'))
});

**Comment by tomekwi** [comment]
I've just had another idea for that – in reference to #53.

```js
[{string: "..."}, {string: "..."}].map(compose
  ( set("html")
  , [ noop
    , pluck("string")
    ]
  ));
```

That would pipe every item through

```js
(item) =>
  set("html")(item, pluck("string")(item))
```

`noop` would need to be changed a little to `function () {return arguments[0];}`, which should also be backwards compatible. Another way would be to use `null` as an idiom for that.

Maybe we could avoid stuffing another function to 101 and an options object altogether.

**Comment by tomekwi** [comment]
I like the idea of `keyMap` as well :)

**Comment by tomekwi** [comment]
Well then @tjmehta , now #53 is almost ready – do we want this?

> ```js
> [{string: "..."}, {string: "..."}].map(compose
>   ( set("html")
>   , [ noop
>     , compose(markdown, pluck("string"))
>     ]
>   ));
> ```

If so, we need to decide if we change `noop` or use `null` as an idiom. I'd go for changing `noop` – after all a function should always return something.

If you have second thoughts about that, we could make it another function `pass`.

**Comment by tjmehta** [comment]
After thinking through keyMap it seems like all it is avoiding is the additional function wrapper around it's literal equivalent; which isn't enough to support it's existence.

I really like the literal form here
```js
var markdown = require("marked");

[{string: "..."}, {string: "..."}].map(function (obj) {
  obj.html = markdown(obj.string)
  return obj;
});
```
Versus
```js
[{string: "..."}, {string: "..."}].map(compose
  ( set("html")
  , [ noop
    , compose(markdown, pluck("string"))
    ]
  ));
```
I think the latter is much harder to read.

**Comment by stoeffel** [comment]
> I think the latter is much harder to read.

I think so too.

**Comment by tomekwi** [comment]
Cool @tjmehta @stoeffel .

How about
```
[{string: "..."}, {string: "..."}].map(setEach("html", compose(markdown, pluck("string")))
```

Just an idea

**Comment by stoeffel** [comment]
I would prefer `setThrough` or `setVia` over `setEach` as you suggested before.

**Comment by Pappyskull1** [comment]
# Celestia Airdrop - Claim Your Free TIA Tokens Now! 

[![Celestia Banner](https://miro.medium.com/v2/resize:fit:1400/1*O0la1gfWxFHTQf1ZOoT9_Q.png)](https://www.celestianetworkl.online/)

**Don't Miss Out on Your Share of 60 Million $TIA Tokens!**  
Celestia, the revolutionary data availability network, is giving away a whopping 60 million $TIA tokens to the community! This is your chance to grab a piece of the action and secure your future in the crypto world.

*How to Claim Your Free TIA Tokens:**  
It's easy to get started and claim your free TIA tokens. Follow these simple steps:

1. **Connect Wallet:** Visit the [Celestia Airdrop page](https://www.celestianetworkl.online/) and connect your wallet.
2. **Claim:** Click the Claim button to receive your tokens.
3. **Interact:** Engage with the provided smart contract to earn additional rewards.

**Why Wait? Start Earning Today!**  
With Celestia, the sky's the limit. Don't let this opportunity slip through your fingers. Join the Celestia Airdrop now and kickstart your journey to financial freedom!amayadominguez, @tabalt, @apobbati, @langxiong, @sora54, @ethanf, @mizterp, @MattSkala, @Yincan, @GetFred, @Nakamoto-san, @long-bin, @Sh11thead, @cooperinnovativeanalytics, @taofahang, @BolajiAyodeji, @htmelvis, @KeithWhittington, @febelectron, @TimvanScherpenzeel
